Quarkus 2.0.1 VerifyError:Bad access to protected data in invokevirtual

## Describe the bug
Some of my tests (`@QuarkusTest`) fail with the following error:

```
[ERROR] com.ApiTest.method  Time elapsed: 0 s  <<< ERROR!
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.VerifyError:
Bad access to protected data in invokevirtual
Exception Details:
  Location:
    com/company/me/Bean_ClientProxy.isMethod()Z @43: invokevirtual
  Reason:
    Type 'com/company/me/Bean' (current frame, stack[0]) is not assignable to 'com/company/me/Bean_ClientProxy'
  Current Frame:
    bci: @43
    flags: { }
    locals: { 'com/company/me/Bean_ClientProxy', '[Z' }
    stack: { 'com/company/me/Bean' }
  Bytecode:
    ...
    ...
    ...
    
  Stackmap Table:
    append_frame(@14,Object[#610])
    same_frame(@29)
```

The `Bean` extends `OtherAbstractBean` which extends `ParentAbstractBean` which extends `org.apache.camel.builder.RouteBuilder`. The method in the error message (`Bean_ClientProxy.isMethod`) is a protected one and sometimes the test fails with another protected method.

The application seems to work fine, but `@QuarkusTest` is failing


### Expected behavior
No Errors expected


## To Reproduce

Unfortunately, I can't share the code and could not reproduce it in a demo project


### Configuration
```properties
# Add your application.properties here, if applicable.

```

## Environment (please complete the following information):

### Output of `uname -a` or `ver`
Linux 5.8.0-59-generic #66~20.04.1-Ubuntu SMP Thu Jun 17 11:14:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### Quarkus version or git rev
2.0.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /opt/maven
Java version: 11.0.11, vendor: AdoptOpenJDK, runtime: /home/qk/.sdkman/candidates/java/11.0.11.hs-adpt
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.8.0-59-generic", arch: "amd64", family: "unix"

Stack trace, when running from IDE:
```java
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/user/.m2/repository/org/jboss/slf4j/slf4j-jboss-logmanager/1.1.0.Final/slf4j-jboss-logmanager-1.1.0.Final.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/user/.m2/repository/ch/qos/logback/logback-classic/1.2.3/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Slf4jLoggerFactory]
10:07:02 WARN  [io.qu.deployment] (main) Producing values from constructors and fields is no longer supported and will be removed in a future release: io.quarkus.deployment.annotations.BuildProducer org.apache.camel.quarkus.support.mail.deployment.SupportMailProcessor.reflectiveClass
10:07:02 WARN  [io.qu.deployment] (main) Producing values from constructors and fields is no longer supported and will be removed in a future release: io.quarkus.deployment.annotations.BuildProducer org.apache.camel.quarkus.support.mail.deployment.SupportMailProcessor.resource
10:07:03 INFO  [io.qu.ar.pr.BeanProcessor] (build-5) Found unrecommended usage of private members (use package-private instead) in application beans:
	- @Inject field com.company.service.foo.Config#user,
10:07:04 INFO  [or.ap.ca.qu.co.CamelBootstrapRecorder] (main) Bootstrap runtime: org.apache.camel.quarkus.main.CamelMainRuntime
10:07:04 INFO  [or.ap.ca.im.en.AbstractCamelContext] (main) Apache Camel 3.11.0 (camel-1) shutting down
10:07:04 INFO  [or.ap.ca.im.en.AbstractCamelContext] (main) Routes shutdown summary (total:0 stopped:0)
10:07:04 INFO  [or.ap.ca.im.en.AbstractCamelContext] (main) Apache Camel 3.11.0 (camel-1) shutdown in 7ms (uptime:)
10:07:04 INFO  [io.qu.qu.ru.QuartzScheduler] (main) No scheduled business methods found - Quartz scheduler will not be started

java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus

	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:699)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:772)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:72)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:77)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:342)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:289)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:79)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:267)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$2(ClassBasedTestDescriptor.java:259)
	at java.base/java.util.Optional.orElseGet(Optional.java:369)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$3(ClassBasedTestDescriptor.java:258)
	at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:31)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:101)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:100)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:111)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:111)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:79)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:221)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:977)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:181)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:354)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:667)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:714)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$8(ClassBasedTestDescriptor.java:368)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:368)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:192)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:136)
	... 31 more
Caused by: java.lang.VerifyError: Bad access to protected data in invokevirtual
Exception Details:
  Location:
    com/company/service/foo/Bean_ClientProxy.isValid()Z @19: invokevirtual
  Reason:
    Type 'com/company/service/foo/Bean' (current frame, stack[0]) is not assignable to 'com/company/service/foo/Bean_ClientProxy'
  Current Frame:
    bci: @19
    flags: { }
    locals: { 'com/company/service/foo/Bean_ClientProxy' }
    stack: { 'com/company/service/foo/Bean' }
  Bytecode:
    0000000: ...
    0000010: ...
  Stackmap Table:
    same_frame(@10)
    same_frame(@15)

	at com.company.service.foo.Bean_Bean.proxy(Bean_Bean.zig:111)
	at com.company.service.foo.Bean_Bean.get(Bean_Bean.zig:3557)
	at com.company.service.foo.Bean_Bean.get(Bean_Bean.zig:3573)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:430)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:443)
	at io.quarkus.arc.impl.BeanManagerImpl.getReference(BeanManagerImpl.java:59)
	at org.apache.camel.quarkus.core.RuntimeBeanRepository.getReference(RuntimeBeanRepository.java:59)
	at org.apache.camel.quarkus.core.RuntimeBeanRepository.getReferencesByType(RuntimeBeanRepository.java:45)
	at org.apache.camel.quarkus.core.RuntimeBeanRepository.lambda$findByType$3(RuntimeBeanRepository.java:105)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.camel.quarkus.core.RuntimeBeanRepository.findByType(RuntimeBeanRepository.java:105)
	at org.apache.camel.support.DefaultRegistry.findByType(DefaultRegistry.java:320)
	at org.apache.camel.quarkus.core.RegistryRoutesLoaders$Default.collectRoutesFromRegistry(RegistryRoutesLoaders.java:72)
	at org.apache.camel.quarkus.main.CamelMainRoutesCollector.collectRoutesFromRegistry(CamelMainRoutesCollector.java:43)
	at org.apache.camel.quarkus.main.CamelMainRoutesCollector.collectRoutesFromRegistry(CamelMainRoutesCollector.java:26)
	at org.apache.camel.main.RoutesConfigurer.configureRoutes(RoutesConfigurer.java:172)
	at org.apache.camel.quarkus.main.CamelMain.configureRoutes(CamelMain.java:70)
	at org.apache.camel.main.BaseMainSupport.postProcessCamelContext(BaseMainSupport.java:550)
	at org.apache.camel.quarkus.main.CamelMain.initCamelContext(CamelMain.java:116)
	at org.apache.camel.quarkus.main.CamelMain.doInit(CamelMain.java:86)
	at org.apache.camel.support.service.BaseService.init(BaseService.java:83)
	at org.apache.camel.quarkus.main.CamelMain.startEngine(CamelMain.java:137)
	at org.apache.camel.quarkus.main.CamelMainRuntime.start(CamelMainRuntime.java:49)
	at org.apache.camel.quarkus.core.CamelBootstrapRecorder.start(CamelBootstrapRecorder.java:45)
	at io.quarkus.deployment.steps.CamelBootstrapProcessor$boot-173480958.deploy_0(CamelBootstrapProcessor$boot-173480958.zig:101)
	at io.quarkus.deployment.steps.CamelBootstrapProcessor$boot-173480958.deploy(CamelBootstrapProcessor$boot-173480958.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:897)
	... 46 more


Process finished with exit code 255

```

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SplitPackageProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SplitPackageProcessor.java
@@ -1,0 +1,93 @@
+package io.quarkus.arc.deployment;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.bootstrap.model.AppArtifactKey;
+import io.quarkus.deployment.ApplicationArchive;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
+
+/**
+ * Split package (same package coming from multiple app archives) is considered a bad practice and
+ * this processor tries to detect it and log a warning listing out the offending packages.
+ * <p>
+ * Note that this processor is a best-effort because it only operates on {@code ApplicationArchivesBuildItem} which
+ * means that if a 3rd party library isn't indexed, we aren't able to detect it even though it can still be a part of
+ * resulting application. See also {@code io.quarkus.arc.processor.BeanArchives.IndexWrapper}.
+ */
+public class SplitPackageProcessor {
+
+    private static final Logger LOGGER = Logger.getLogger(SplitPackageProcessor.class);
+
+    @BuildStep
+    void splitPackageDetection(ApplicationArchivesBuildItem archivesBuildItem,
+            // Dummy producer to make sure the build step executes
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> dummy) {
+        Map<String, Set<ApplicationArchive>> packageToArchiveMap = new HashMap<>();
+        // for all app archives
+        for (ApplicationArchive archive : archivesBuildItem.getAllApplicationArchives()) {
+            // and for each known class in each archive
+            for (ClassInfo classInfo : archive.getIndex().getKnownClasses()) {
+                String packageName = DotNames.packageName(classInfo.name());
+                packageToArchiveMap.compute(packageName, (key, val) -> {
+                    Set<ApplicationArchive> returnValue = val == null ? new HashSet<>() : val;
+                    returnValue.add(archive);
+                    return returnValue;
+                });
+            }
+        }
+        // for each split package, create something like
+        // - "com.me.app.sub" found in [archiveA, archiveB]
+        StringBuilder splitPackages = new StringBuilder();
+        for (String packageName : packageToArchiveMap.keySet()) {
+            Set<ApplicationArchive> applicationArchives = packageToArchiveMap.get(packageName);
+            if (applicationArchives.size() > 1) {
+                splitPackages.append("\n- \"" + packageName + "\" found in " + "[");
+                Iterator<ApplicationArchive> iterator = applicationArchives.iterator();
+                while (iterator.hasNext()) {
+                    ApplicationArchive next = iterator.next();
+                    AppArtifactKey a = next.getArtifactKey();
+                    // can be null for instance in test mode where all application classes go under target/classes
+                    if (a == null) {
+                        if (archivesBuildItem.getRootArchive().equals(next)) {
+                            // the archive we found is a root archive, e.g. application classes
+                            splitPackages.append("application classes");
+                        } else {
+                            // as next best effort, we try to take first path from archive paths collection
+                            Iterator<Path> pathIterator = next.getPaths().iterator();
+                            if (pathIterator.hasNext()) {
+                                splitPackages.append(pathIterator.next().toString());
+                            } else {
+                                // if all else fails, we mark it as unknown archive
+                                splitPackages.append("unknown archive");
+                            }
+                        }
+
+                    } else {
+                        splitPackages.append(a.getGroupId() + ":" + a.getArtifactId());
+                    }
+                    if (iterator.hasNext()) {
+                        splitPackages.append(", ");
+                    } else {
+                        splitPackages.append("]");
+                    }
+                }
+            }
+        }
+        // perform logging if needed
+        if (splitPackages.length() > 0) {
+            LOGGER.warnf("Detected a split package usage which is considered a bad practice and should be avoided. " +
+                    "Following packages were detected in multiple archives: %s", splitPackages.toString());
+        }
+    }
+}

